aesd-char-driver/aesd-circular-buffer.c

```diff
@@ -29,9 +30,35 @@
 struct aesd_buffer_entry *aesd_circular_buffer_find_entry_offset_for_fpos(struct aesd_circular_buffer *buffer,
             size_t char_offset, size_t *entry_offset_byte_rtn )
 {
-    /**
-    * TODO: implement per description
-    */
+    size_t total_offset = 0;
+    size_t entry_size;
+    struct aesd_buffer_entry *entry;
+    if (buffer == NULL || entry_offset_byte_rtn == NULL) {
+        return NULL;
+    }
+    // Iterate over all the valid entries in the buffer
+    for (int i = 0; i < AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED; i++) {
+        int index = (buffer->out_offs + i) % AESDCHAR_MAX_WRITE_OPERATIONS_SUPPORTED;
+        
+        entry = &buffer->entry[index];
+        entry_size = entry->size;
+        
+        // Check if char_offset is within the current entry
+        if (char_offset < total_offset + entry_size) {
+            // If found, calculate the offset within the entry
+            *entry_offset_byte_rtn = char_offset - total_offset;
+            return entry;
+        }
+        
+        total_offset += entry_size;
+        
+        // Break out if we've wrapped around and are at the starting point
+        if (index == buffer->in_offs && !buffer->full) {
```

Comment:
@pranjalgupta1901 I don't think you need the `!buffer->full` check here.  Anytime you hit the input pointer (or pass it) you will be looking at old circular buffer entries.  I think this logic works because we never transition back out of full state once in it.